File: app/src/lib/photo-index-repository.ts
export interface PhotoDoc {
  googlePhotosId: string;
  filename: string;
  takenAt: Date;
  previewGcsPath: string;
  originalGcsPath: string;
  width: number;
  height: number;
  latitude: number | null;
  longitude: number | null;
}

export interface PhotoResult {
  id: string;
  takenAt: Date;
  previewGcsPath: string | null;
  originalGcsPath: string;
  width: number | null;
  height: number | null;
}

export interface ListOptions {
  limit: number;
  cursor?: string;
}

export interface ListResult {
  photos: PhotoResult[];
  nextCursor: string | null;
}

export class PhotoIndexRepository {
  constructor(private db: FirebaseFirestore.Firestore) {}

  async list({ limit, cursor }: ListOptions): Promise<ListResult> {
    let query = this.db
      .collection("photos")
      .orderBy("taken_at", "desc")
      .limit(limit);

    if (cursor) {
      const cursorDoc = await this.db.collection("photos").doc(cursor).get();
      if (cursorDoc.exists) {
        query = query.startAfter(cursorDoc);
      }
    }

    const snapshot = await query.get();
    const photos: PhotoResult[] = snapshot.docs.map((doc) => {
      const data = doc.data();
      return {
        id: doc.id,
        takenAt: data.taken_at.toDate(),
        previewGcsPath: data.preview_gcs_path ?? null,
        originalGcsPath: data.original_gcs_path,
        width: data.width ?? null,
        height: data.height ?? null,
      };
    });

    const nextCursor =
      snapshot.docs.length === limit
        ? snapshot.docs[snapshot.docs.length - 1].id
        : null;

    return { photos, nextCursor };
  }
}

File: jobs/ingest/src/photo_index_repository.py
from dataclasses import dataclass
from datetime import datetime
from typing import Optional


@dataclass
class PhotoDoc:
    google_photos_id: str
    filename: str
    taken_at: datetime
    preview_gcs_path: str
    original_gcs_path: str
    width: int
    height: int
    latitude: Optional[float]
    longitude: Optional[float]


class PhotoIndexRepository:
    COLLECTION = "photos"

    def __init__(self, db):
        self._db = db

    def exists(self, google_photos_id: str) -> bool:
        results = (
            self._db.collection(self.COLLECTION)
            .where("google_photos_id", "==", google_photos_id)
            .limit(1)
            .get()
        )
        return len(results) > 0

    def upsert(self, doc: PhotoDoc) -> None:
        data = {
            "google_photos_id": doc.google_photos_id,
            "filename": doc.filename,
            "taken_at": doc.taken_at,
            "preview_gcs_path": doc.preview_gcs_path,
            "original_gcs_path": doc.original_gcs_path,
            "width": doc.width,
            "height": doc.height,
            "latitude": doc.latitude,
            "longitude": doc.longitude,
        }
        self._db.collection(self.COLLECTION).document(doc.google_photos_id).set(data)
